src/impls.rs

```diff
@@ -1175,18 +1128,55 @@ impl<T: Config> Pallet<T> {
 	/// # Arguments
 	/// * `query_id` - Identifier of the specific data feed.
 	/// * `timestamp` - The timestamp of the value to remove.
-	pub(super) fn remove_value(query_id: QueryId, timestamp: Timestamp) -> DispatchResult {
-		ensure!(!<ReportDisputes<T>>::get(query_id, timestamp), Error::<T>::ValueDisputed);
-		let index = <ReportedTimestamps<T>>::get(query_id, timestamp)
-			.ok_or(Error::<T>::InvalidTimestamp)?;
-		ensure!(
-			Some(timestamp) == <ReportedTimestampsByIndex<T>>::get(query_id, index),
-			Error::<T>::InvalidTimestamp
-		);
+	pub(super) fn remove_value(
+		query_id: QueryId,
+		timestamp: Timestamp,
+	) -> Result<u32, DispatchError> {
+		let iterations =
+			<Reports<T>>::try_mutate(query_id, timestamp, |maybe| -> Result<u32, DispatchError> {
+				let Some(report) = maybe else { return Err(Error::<T>::InvalidTimestamp.into()) };
+				ensure!(!report.is_disputed, Error::<T>::ValueDisputed);
+				ensure!(
+					Some(timestamp) == <ReportedTimestampsByIndex<T>>::get(query_id, report.index),
+					Error::<T>::InvalidTimestamp
+				);
+				report.is_disputed = true;
+
+				// Update last reported timestamp, if applicable
+				let _ = <LastReportedTimestamp<T>>::try_mutate(query_id, |lrt| match lrt {
+					// Check if last reported timestamp value is being removed
+					Some(last_reported_timestamp) if *last_reported_timestamp == timestamp => {
+						*lrt = report.previous; // set last reported to previous
+						return Ok(())
+					},
+					_ => Err(()), // No mutation
+				});
+
+				// Update next valid timestamp in series to point to previous valid timestamp (before one being removed)
+				let start = report.index.checked_add(1).ok_or(ArithmeticError::Overflow)?;
+				let end =
```

Comment:
can't we get the total report submissions for particular query_id? With this we don't need to rely on extra config constant and we'll get the exact length as well?
I agree, it'll get one extra db read but we'll get exact records.

Reply:
We need some constant to bound the number of iterations, as using u32::max would make the extrinsic unusable as found currently with tip().

The config item basically allows specification of the maximum number of disputed timestamps in a row that it will check before it errors.

I guess a simpler approach might be to use the total reports as the range end as you suggest and then check if iteration count reaches the max configured amount before erroring.

Reply:
I simplified the max iterations logic, which helped address the other range issue you pointed out. Please do let me know if you still see any issues.

Reply:
Looks good,